llvm/lib/CodeGen/MachineRegisterInfo.cpp

```diff
@@ -650,6 +654,31 @@ void MachineRegisterInfo::setCalleeSavedRegs(ArrayRef<MCPhysReg> CSRs) {
   IsUpdatedCSRsInitialized = true;
 }
 
+void MachineRegisterInfo::initializeRegClassSyntheticInfo() {
+  const TargetRegisterInfo *TRI = getTargetRegisterInfo();
+
+  RegClassSyntheticInfo.resize(TRI->getNumRegClasses());
+  for (const TargetRegisterClass *RC : TRI->regclasses()) {
+    if (RC->isSynthetic())
+      RegClassSyntheticInfo.set(RC->getID());
+  }
+}
+
+void MachineRegisterInfo::changeSyntheticInfoForRC(
+    const TargetRegisterClass *RC, bool Value) {
+  assert(RC->isSynthetic() && "Regclasses can be enabled/disabled dynamically "
+                              "only if marked synthetic.");
+
+  if (Value)
+    RegClassSyntheticInfo.set(RC->getID());
+  else
+    RegClassSyntheticInfo.reset(RC->getID());
+}
+
+bool MachineRegisterInfo::isEnabled(const TargetRegisterClass *RC) const {
+  return !RegClassSyntheticInfo.test(RC->getID());
+}
```

Comment:
These functions feel out of place in this PR, no one is calling them, they aren't being tested. I haven't looked at the rest of the PRs yet. 1) Can you tell more about how they will be used? 2) Can there be unittests for them here?

Reply:
This patch sets up the whole machinery for the new field Synthetic that I'm introducing.
The test llvm/unittests/CodeGen/MachineInstrTest.cpp covers it. The helper function `isEnabled`, is used in the regclass queries in MRI and TRI interfaces in this patch itself.